fdbserver/workloads/MetaclusterManagementWorkload.actor.cpp

```diff
@@ -272,40 +296,155 @@ struct MetaclusterManagementWorkload : TestWorkload {
+				}
+
+				wait(waitForAll(getFutures));
+
+				state std::vector<Future<Void>> deleteFutures;
+				for (auto const& f : getFutures) {
+					ASSERT(f.get().present());
+					deleteFutures.push_back(TenantAPI::deleteTenantTransaction(tr, f.get().get()));
+				}
+
+				wait(waitForAll(deleteFutures));
+				wait(tr->commit());
+				break;
+			} catch (Error& e) {
+				wait(tr->onError(e));
+			}
+		}
+
+		ASSERT(foundTenantCollision || foundGroupCollision);
+		return Void();
+	}
+
 	ACTOR static Future<Void> restoreCluster(MetaclusterManagementWorkload* self) {
 		state ClusterName clusterName = self->chooseClusterName();
-		state DataClusterData* dataDb = &self->dataDbs[clusterName];
+		state Reference<DataClusterData> dataDb = self->dataDbs[clusterName];
 		state bool dryRun = deterministicRandom()->coinflip();
 		state bool forceJoin = deterministicRandom()->coinflip();
 
 		state std::vector<std::string> messages;
 		state bool retried = false;
 		loop {
 			try {
+				if (dataDb->detached) {
+					if (retried) {
+						wait(removeFailedRestoredCluster(self, clusterName));
+					} else {
+						// On the first try, we need to remove the metacluster registration entry from the data
+						// cluster
+						wait(success(MetaclusterAPI::removeCluster(
+						    dataDb->db.getReference(), clusterName, ClusterType::METACLUSTER_DATA, ForceRemove::True)));
+					}
+				}
+
 				Future<Void> restoreFuture =
 				    MetaclusterAPI::restoreCluster(self->managementDb,
 				                                   clusterName,
 				                                   dataDb->db->getConnectionRecord()->getConnectionString(),
-				                                   ApplyManagementClusterUpdates::True,
+				                                   ApplyManagementClusterUpdates(!dataDb->detached),
 				                                   RestoreDryRun(dryRun),
 				                                   ForceJoin(forceJoin),
 				                                   &messages);
 				Optional<Void> result = wait(timeout(restoreFuture, deterministicRandom()->randomInt(1, 30)));
-				if (result.present()) {
-					break;
+				if (!result.present()) {
+					retried = true;
+					continue;
+				}
+
+				ASSERT(dataDb->registered || dataDb->detached);
+				if (dataDb->detached && !dryRun) {
+					dataDb->detached = false;
+					dataDb->registered = true;
+					for (auto const& t : dataDb->ungroupedTenants) {
+						self->ungroupedTenants.insert(t);
+					}
+					for (auto const& t : dataDb->tenantGroups) {
+						ASSERT(self->tenantGroups.try_emplace(t.first, t.second).second);
+					}
+					for (auto const& t : dataDb->tenants) {
+						ASSERT(self->createdTenants.try_emplace(t.first, t.second).second);
+						ASSERT(self->createdTenants[t.first]->cluster == clusterName);
+					}
+
+					self->totalTenantGroupCapacity += dataDb->ungroupedTenants.size() + dataDb->tenantGroups.size();
 				}
-				retried = true;
+
+				break;
 			} catch (Error& e) {
+				state Error error = e;
```

Comment:
Same comment about mixing usage of `error` and `e`

Reply:
This one seems like it's probably ok. I'm stashing the error for use in the final trace event, but all usages prior to that consistently use `e`.

Reply:
I decided to change it to match the other, so it uses `error` consistently now